pulley/src/profile.rs

```diff
@@ -0,0 +1,248 @@
+//! Low-level support for profiling pulley.
+//!
+//! This is used in conjunction with the `profiler-html.rs` example with Pulley
+//! and the `pulley.rs` ProfilingAgent in Wasmtime.
+
+use anyhow::{anyhow, bail, Context, Result};
+use std::fs::{File, OpenOptions};
+use std::io::{BufWriter, Write};
+use std::sync::atomic::{AtomicUsize, Ordering::Relaxed};
+use std::sync::Arc;
+use std::vec::Vec;
+
+// Header markers for sections in the binary `*.data` file.
+
+/// Section of the `*.data` file which looks like:
+///
+/// ```text
+/// * byte: ID_FUNCTION
+/// * addr: 8-byte little-endian address that this body was located at
+/// * name_len: 4-byte little-endian byte length of `name`
+/// * name: contents of the name of the function
+/// * body_len: 4-byte little-endian byte length of `body`
+/// * body: contents of the body of the function
+/// ```
+const ID_FUNCTION: u8 = 1;
+
+/// Section of the `*.data` file which looks like:
+///
+/// ```text
+/// * byte: ID_SAMPLES
+/// * sample_len: 4-byte little-endian element count of `samples`
+/// * samples: sequence of 8-byte little endian addresses
+/// ```
+const ID_SAMPLES: u8 = 2;
+
+/// Representation of a currently executing program counter of an interpreter.
+///
+/// Stores an `Arc` internally that is safe to clone/read from other threads.
+#[derive(Default, Clone)]
+pub struct ExecutingPc(Arc<AtomicUsize>);
+
+impl ExecutingPc {
+    pub(crate) fn as_ref(&self) -> ExecutingPcRef<'_> {
+        ExecutingPcRef(&self.0)
+    }
+
+    /// Loads the currently executing program counter, if the interpreter is
+    /// running.
+    pub fn get(&self) -> Option<usize> {
+        match self.0.load(Relaxed) {
+            0 => None,
+            n => Some(n),
+        }
+    }
+
+    /// Returns whether the interpreter has been destroyed and will no longer
+    /// execute any code.
+    pub fn is_done(&self) -> bool {
+        // TODO: this doesn't work with `Clone` above in general and only works
+        // given the strict pattern of how everything is integrated today.
+        // Should probably integrate an `AtomicBool` or something like that and
+        // hook into `Drop for Vm` or something like that.
```

Comment:
This seems like a fairly trivial change in terms of effort, but good payoff in terms of longer-term maintainability. Mind doing it now?